IntegrationTests/UI.IntegrationTests/UITest.cs

```diff
@@ -27,32 +28,69 @@ void HandleApplicationIdle(object sender, EventArgs e)
         }
 
         public static void RunForm<T>(
-            Action showDialog,
-            Func<T, Task> runAsync)
+            Action showForm,
+            Func<T, Task> runTestAsync)
             where T : Form
         {
-            // Avoid using ThreadHelper.JoinableTaskFactory for the outermost operation because we don't want the task
-            // tracked by its collection. Otherwise, test code would not be able to wait for pending operations to
-            // complete.
-            var test = ThreadHelper.JoinableTaskContext.Factory.RunAsync(async () =>
+            // If form.Dipose was called by the async test task, closing the window would be done in a strange order.
+            T form = null;
+            try
             {
-                await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
-                await WaitForIdleAsync();
-                var dialog = Application.OpenForms.OfType<T>().Single();
-                try
-                {
-                    await runAsync(dialog);
-                }
-                finally
+                // Start runTestAsync before calling showForm.
+                // The latter might block until the form is closed, especially if using Application.Run(form).
+
+                // Avoid using ThreadHelper.JoinableTaskFactory for the outermost operation because we don't want the task
+                // tracked by its collection. Otherwise, test code would not be able to wait for pending operations to
+                // complete.
+                var test = ThreadHelper.JoinableTaskContext.Factory.RunAsync(async () =>
                 {
-                    dialog.Close();
-                }
-            });
+                    try
+                    {
+                        await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+                        await WaitForIdleAsync();
+                        form = Application.OpenForms.OfType<T>().Single();
```

Comment:
📝 These lines do not need to be in the `try` block

Reply:
> These lines do not need to be in the try block.

They really should. They or added logging code could throw an exception, which results in a hang because no one closes the tested form. Hence `Application.Run(form)` and `showForm` never return.

Reply:
> They or added logging code could throw an exception

Exceptions can only be thrown here if:

* There is no main thread
* There is no form of the specific type open
* There is more than one form of the specified type open

For the first two cases, there is no cleanup necessary. For the third case, I would prefer to only take action if it's been a demonstrated problem, and even then I would prefer to assert at the very start of the test that there are no forms of the specified type open.

Reply:
You're right regarding "the earlier try does not help".
But such an exception is not reported. The test just hangs with an open form.
That's why I tend to add another `try catch(ex) { Console.WriteLine(ex)`.
(When I added debug logging, the inserted catch did not catch a NRE from an early attempt to log something copied & pasted.)

> assert at the very start of the test that there are no forms of the specified type open

👍

Reply:
Is it possible to reproduce this locally?

Reply:
Nothing easier than this 😉
```
           var test = ThreadHelper.JoinableTaskContext.Factory.RunAsync(async () =>
            {
throw new Exception("let it hang");
                await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
```

Reply:
That undermines my claim that those three lines cannot throw a problematic exception.

I found two sources of potential issues during testing (~800 iterations so far). Will submit a small PR shortly.

Reply:
Are you still analyzing, @sharwell? Are there further review comments?
Else this PR could be squash-merged.

Reply:
I'm looking now

Reply:
I'll send a PR to your fork so you can see my full proposed changes.

Reply:
https://github.com/mstv/gitextensions/pull/1

Reply:
Merged although I disagree with the error handling strategy: https://github.com/mstv/gitextensions/pull/1#discussion_r380965921, https://github.com/mstv/gitextensions/pull/1#discussion_r380967048.